src/DidCache.ts

```diff
@@ -1,16 +1,338 @@
+  readonly batchFileHash: string;
+  readonly type: OperationType;
+}
+
+/**
+ * DIDCache is a singleton class whose instance holds most of the state in Sidetree node.
+ * It exposes methods to record sidetree DID state changes (create, update, delete, recover)
+ * and methods to retrieve current and historical states of a DID document.
+ *
+ * The current implementation is a main-memory implementation without any persistence. This
+ * means that when a node is powered down and restarted DID operations need to be applied
+ * from the beginning of time. This implementation will be extended in the future to support
+ * persistence.
+ */
+export class DidCache {
+  /**
+   * Map a versionId to the next versionId whenever one exists.
+   */
+  private nextVersion: Map<VersionId, VersionId> = new Map();
+
+  /**
+   * Map a operation hash to the OperationInfo which contains sufficient
+   * information to reconstruct the operation.
+   */
+  private opHashToInfo: Map<OperationHash, OperationInfo> = new Map();
+
+  /**
+   * Apply (perform) a specified DID state changing operation.
+   */
+  public apply (operation: WriteOperation): string | null {
+    const opHash = DidCache.getHash(operation);
+
+    // Ignore operations without the required metadata - any operation anchored
+    // in a blockchain should have this metadata.
+    if (operation.transactionNumber === undefined ||
+        operation.operationIndex === undefined ||
+        operation.batchFileHash === undefined) {
+      return null;
+    }
+
+    // opInfo is operation with derivable properties projected out
+    const opInfo: OperationInfo = {
+      transactionNumber: operation.transactionNumber,
+      operationIndex: operation.operationIndex,
+      batchFileHash: operation.batchFileHash,
+      type: operation.type
+    };
+
+    // If this is a duplicate of an earlier operation, we can
+    // ignore this operation. Note that we might have a previous
+    // operation with the same hash, but that previous operation
+    // need not be earlier in timestamp order - hence the check
+    // with lesser().
+    const prevOperation = this.opHashToInfo.get(opHash);
+    if (prevOperation !== undefined && lesser(prevOperation, opInfo)) {
+      return null;
+    }
+    // Update our mapping of operation hash to operation info overwriting
+    // previous info if it exists
+    this.opHashToInfo.set(opHash, opInfo);
+
+    // For operations that have a previous version, we need additional
+    // bookkeeping
+    if (operation.previousOperationHash) {
+      this.applyOpWithPrev(opHash, opInfo, operation.previousOperationHash);
+    }
+
+    return opHash;
+  }
+
+  public constructor (
+    private readonly cas: Cas,
+    private readonly didDocUpdate: DidDocumentUpdater,
+    private readonly didDocCreate: DidDocumentCreator) {
+
+  }
+
+  /**
+   * Rollback the state of the DidCache by removing all operations
+   * with transactionNumber greater than the provided parameter value.
```

Comment:
I think the definition and implementation of `rollback` needs to change to:

_"... greater than OR EQUAL to the provided parameter value."_

It doesn't make sense to say "rollbase X", but it actually mean roll back AFTER X.

Reply:
I took your suggestion.

Reply:
I updated method summary and implementation doc accordingly.